File: usecase/core/src/main/java/com/hearlers/gateway/factory/DefaultOAuthProviderFactory.java
package com.hearlers.gateway.factory;

import com.hearlers.api.proto.v1.model.AuthChannel;
import com.hearlers.com.hearlers.gateway.port.OAuthProviderPort;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

@Component
public class DefaultOAuthProviderFactory implements OAuthProviderFactory {
    private final Map<AuthChannel, OAuthProviderPort> providerMap;


    public DefaultOAuthProviderFactory(List<OAuthProviderPort> providers) {
        this.providerMap = providers.stream()
                .collect(Collectors.toMap(
                        OAuthProviderPort::getSupportedChannel,
                        Function.identity()
                ));
    }
    
    /**
     * 인증 채널에 맞는 OAuth 제공자 서비스를 반환합니다.
     */
    @Override
    public OAuthProviderPort getOAuthProviderPort(AuthChannel authChannel) {
        OAuthProviderPort provider = providerMap.get(authChannel);
        if (provider == null) {
            throw new IllegalArgumentException("Unsupported auth channel: " + authChannel);
        }
        return provider;
    }
}

File: usecase/core/src/main/kotlin/com/hearlers/gateway/factory/DefaultOAuthProviderFactory.kt
package com.hearlers.gateway.factory

import com.hearlers.api.proto.v1.model.AuthChannel
import com.hearlers.gateway.port.OAuthProviderPort

import org.springframework.stereotype.Component
import java.util.EnumMap

@Component
class DefaultOAuthProviderFactory(
    providers: List<OAuthProviderPort>
) : OAuthProviderFactory {

    private val providerMap: Map<AuthChannel, OAuthProviderPort> =
        EnumMap(providers.associateBy { it.supportedChannel })

    override fun getOAuthProviderPort(authChannel: AuthChannel): OAuthProviderPort {
        return providerMap[authChannel]
            ?: throw IllegalArgumentException("지원하지 않는 인증 채널입니다: $authChannel")
    }
}